Step 1.1: 
Can you think of a reason why this sample does not match the expected GC content distribution and base content at the beginning of sequences?
Ans: The analysis is meant for reads from the whole genome, but we are looking at reads corresponding to mRNA transcripts, so the specific mRNA in A1 may have a different GC content relative to the genome. The base content at the beginning may correspond to the polyA tail found on all mRNAs.

Step 1.2: 
What is the origin of the most overrepresented sequence in this sample? Does it make sense?
Ans: Serine protease. This makes sense because it is a sample from the fly gut, where proteases will be expressed.

Step 2:
If you were to reject any samples with the percentage of unique reads less than 45%, how many samples would you keep?
Ans: I would keep 25 samples out of the 30 as they are above the 45% cutoff.

Can you see the blocks of triplicates clearly? Try adjusting the min slider up. Does this suggest anything about consistency between replicates?
Ans: I can see the triplicates for some samples clearly (a dark 3x3 square centered on the diagonal). However, there are some samples where this square is not so easy to see, suggesting high variation between replicates. 

Step 3.3:
What does the PCA plot suggest to you about the data? Why?
Ans: Because there is a mismatch in clustering between one LFC-Fe and Fe replicate it suggests these may have been mislabeled during the experiment and sequencing. 

Step 3.6:
Do the categories of enrichments make sense? Why?
Ans: They make sense, most are genes in catabolic and anabolic pathways. Because this is the gut, I would expect processes for digestion and biosynthesis of food.


